Validate RPC log/callTrace response

khaidarkairbek opened the pull request:
Added a check to validate whether filtered logs and calltraces point to valid transactions in the block for realtime-sync engine. Relates to #1150 issue.

khaidarkairbek commented:
@kyscott18 Added a check after filtering transactions, was wondering if this would be sufficient for realtime-sync

0xOlias commented on packages/core/src/sync-realtime/index.ts:
```diff
@@ -497,6 +497,13 @@ export const createRealtimeSync = (
 
+    // Validate that filtered logs/callTraces point to valid transaction in the block
+    if (transactions.length !== requiredTransactions.size) {
+      throw new Error(
+        `Received callTrace/log in the block referencing transaction that is not in current head block '${block.hash}'`,
+      );
```
Not blocking, but I'd advocate for a more specific error message which includes the suspicious transaction hashes. That would make it much easier for us to submit bug reports to chains / RPC providers.

Also, what if the lengths are different because we have an _extra_ required transaction? That scenario would suggest a Ponder bug rather than a faulty RPC.

khaidarkairbek replied:
1) With regards to error message, do you think it would be better to include the list of suspicious tx hashes in the message? 
2) We would expect requiredTransactions size to be larger than transactions length only in the case if some logs or callTraces that is matched with filter are pointing to more transactions, than there are transactions filtered by those logs/traces. 
`const transactions = block.transactions.filter(({ hash }) =>
      requiredTransactions.has(hash),
    );`

kyscott18 replied:
1. yes
2. I think the transaction validation should be more precise. Something like:
```ts
const blockTransactionsHashes = new Set(
  block.transactions.map((t) => t.hash),
);
for (const hash of Array.from(requiredTransactions)) {
  if (blockTransactionsHashes.has(hash) === false) {
    // throw error, include `hash`
  }
}
```

I also think it would be nice to have something that compares `blockHash` in `logs` and `callTraces` to `block.hash`

khaidarkairbek commented:
@kyscott18 @0xOlias I have added the validation checks in sync-realtime for logs response: 
 - log.transactionHash in block.transactions
 - log.blockHash === block.hash
and for callTraces response: 
 - trace.transactionHash in block.transactions
 - trace.blockHash === block.hash (already existed)
 
Similarly, for sync-historical I have added the following checks: 
 - log.transactionHash in block.transactions
 - trace.transactionHash in block.transactions
 
Since logs and traces are fetched by batches in the sync-historical, it's not feasible to check if the logs and traces returned by the rpc point to the right block. I suggest adding the check if the block that log/trace point to is in the range of blocks fetched from rpc.

kyscott18 commented on packages/core/src/sync-realtime/index.ts:
```diff
@@ -393,6 +393,15 @@ export const createRealtimeSync = (
       }
+
+      // Check that logs refer to the correct block
+      for (const log of logs) {
```
This is great
